core/blockchain.go

```diff
@@ -3444,36 +3465,86 @@ func (bc *BlockChain) PreviousCanonicalCoincidentOnPath(header *types.Header, sl
 // dominant block is canonical.
 func (bc *BlockChain) CheckCanonical(header *types.Header, order int) error {
 	lastUncleHeader := &types.Header{}
+	lastUncleHash := common.Hash{}
 	for {
 		status := bc.domClient.GetBlockStatus(context.Background(), header)
 		// If the header is cononical break else keep looking
 		switch status {
 		case quaiclient.CanonStatTy:
-			if (lastUncleHeader != &types.Header{}) {
+			if (lastUncleHash != common.Hash{}) {
+				fmt.Println("header", header.Number, header.Hash(), lastUncleHeader.Number, lastUncleHash)
 				bc.ReOrgRollBack(lastUncleHeader, []*types.Header{}, []*types.Header{})
+				return consensus.ErrSubordinateNotSynced
 			}
 			return nil
 		default:
 			lastUncleHeader = header
+			lastUncleHash = lastUncleHeader.Hash()
 		}
 
 		if header.Number[types.QuaiNetworkContext].Cmp(big.NewInt(0)) == 0 {
+			if (lastUncleHeader != &types.Header{}) {
+				return consensus.ErrSubordinateNotSynced
+			}
 			return nil
 		}
 
 		terminalHeader, err := bc.Engine().PreviousCoincidentOnPath(bc, header, header.Location, order, types.QuaiNetworkContext, true)
 		if err != nil {
+			log.Warn("Error in PCOP", "err", err)
 			return err
 		}
 
 		if terminalHeader.Number[types.QuaiNetworkContext].Cmp(big.NewInt(0)) == 0 {
+			if (lastUncleHeader != &types.Header{}) {
+				return consensus.ErrSubordinateNotSynced
+			}
 			return nil
 		}
 
 		header = terminalHeader
 	}
 }
 
+// CheckSubordinateHeader asks the subordinate chain what their header is and will manifest
+// the subordinate chain to the proper head.
+func (bc *BlockChain) CheckSubordinateHeader(header *types.Header) error {
```

Comment:
This function is misleading. By name alone, I would assume this function asked its sub to perform some set of checks/validation on the header. Instead, it looks like this function just checks this block is a simple extension of the subordinate head (and if not, sync the sub). Maybe rename to `ConfirmHeaderExtendsSubordinate()` or something like that?